### ISSUE TITLE
PHP and HTML

### ISSUE THREAD

**Issue by dianakleban** [susanBuck/dwa15-fall2017#16]
I am new to PHP and struggling to understand the best ways to embed php inside html. I read the notes in github too. Let's say I want to I want to create everything in the same file **without** using "include". I am confused why it is ok to use it in the first example:

![image](https://user-images.githubusercontent.com/31451809/30303459-e1797d92-972d-11e7-8d46-4b4edcfa95a9.png)

----------------------------------------------------------
But it is NOT ok to do this:
![image](https://user-images.githubusercontent.com/31451809/30303471-f5f4ec0c-972d-11e7-841c-dc192f8916bb.png)

Logic on top, display on the bottom. What am I missing?

Can we embed PHP inside HTML like this? What are the best practices and what should we use for the first Project1.

![image](https://user-images.githubusercontent.com/31451809/30303434-c494bf52-972d-11e7-8df2-161a6a2bda1c.png)

**Comment by sheetaldalvi** [comment]
As I understand, the good practice is to keep the business logic and display separate as it helps maintain the code better. Also, the "island" code scattered in the single file is difficult to read and not considered good design. I would rather have the business logic in  different php file.

**Comment by HeinrichTremblay** [comment]
PHP can be embed in HTML just like javascript or CSS can. However, as your codebase grows, it is good practise to separate your code logic (ex: the random quote generator) from the presentation (ex: the html code). Imagine you have a very complex web app with thousand of lines, if you keep everything together it will become a nightmare.

Another exemple to understand this concept would be your CSS, you keep it in an `.css` file instead of in your `.html` file, for the sake of separating content from styling and not ending up with a giant file. 

**Short answer:** You have to separate your logic into it's own `.php` file that you include in top of your index.php file.


```php
<?php require 'your-logic.php'; ?>
<!DOCTYPE html>
<html lang="en">
...
```

**Comment by dianakleban** [comment]
For the sake of simplicity for project1, I don't want to include/require any extra files. I would like to understand how to put PHP and HTML together in the same file.

**Comment by zeroender** [comment]
Heinrich and others pretty much covered it, but I'll throw in my 2 cents.  

a php file can, for all intents and purposes, include everything a html file can.  You could take any html file and rename it to have .php as the extension, and it would work exactly the same.  Inside of the file, as you can see in the sections, the places for php, html, etc code is defined.

Functionally, there is no difference between having all your code be "inline" in html elements - php, css, etc can be declared and used inside the definition of an html element.  From the interpreter/browser point of view, there isn't a difference.  

The difference comes in when you think about how to update one section without changing the other, which comes into play when you have bigger applications or when more than one person is working on the same files.  In the section where one defines the page structure - which is html - you want to only be thinking about how the page should be organized, and not have to consider php logic or css values.  Likewise, when you're thinking about how to generate a random quote, you don't want to think about page placement or font values - Keep It Simple!

In the example above, there is PHP in both the top section - which is all PHP code - and in the bottom, which is html code.  However, the PHP in the html section is just an echo statement - it's just displaying a value.  This is the absolute minimum php you can have to display a value that you computed or defined in the PHP only section.  There is more you can look at with php/html templating, but avoid that for now. 

PHP in one section
HTML in the other, with nothing more than a PHP echo statement.  No "brains" in the PHP code here - just repeating something defined elsewhere.

Hopefully that's helpful!

Tl;dr: 
1. Define a PHP section on top, close it.  
2. Define a HTML section underneath, create all the tags, and in one block (I used a paragraph), instead of text, have a php echo statement, wrapped in the required PHP begin and end blocks.  I wrote my page with text in the "quote" section, then replaced it at the end.

**Comment by ALiborio** [comment]
> For the sake of simplicity for project1, I don't want to include/require any extra files. I would like to understand how to put PHP and HTML together in the same file.

The example you asked about is valid, and it will function the same way. When I originally wrote my PHP logic for project 1 I did it all in the main file right in the spot I wanted to display my random quote. Once I was pleased with how it was functioning I went back and refactored it to move the logic into another file as suggested in the notes. I then only needed the display logic embedded within the HTML tags to display the actual quote.

So putting all your logic embedded within the HTML which will display it will function but it is not the best practice as it gets unwieldy as the web application gets larger. Putting it all at the top of the file or in a separate file makes it easier to maintain and the HTML structure will be less cluttered with PHP code.

**Comment by zeroender** [comment]
Adding on again :)

Even if you have all your html and PHP code in the same place, it should appear as seperate chunks, with the one exception of the echo statements.  At that point, the chunk of PHP code could be cut and pasted into a seperate file - eg "code.php" - and included at the top of the index file.  Once the code is seperate in this manner, it's both functionally and logically seperate.

Another way to approach this would be to create just the website without the quote and a placeholder, then create the PHP and have the quote echo'd in the PHP section (it will probally appear on the top of the page).  Then you'll have two seperate sections for PHP and HTML, and you can just cut and paste the echo into the html section (including the php tags).

HTH!

**Comment by luke777gu** [comment]
Thanks for the reminder. Actually, I did not realize that used the bad way to write my php until I read this issue post. I will also link one webpage with all of the best practices. 
@http://flowframework.readthedocs.io/en/stable/TheDefinitiveGuide/PartV/CodingGuideLines/PHP.html

**Comment by ysrinu** [comment]
dianakleban, understand you don't want to create new php file for the sake of simplicity. This assignment might also be the best place to practice separating out the php logic code into a separate file. It's much easier to debug and fix the separation of logic and display in this example.

**Comment by peggyo** [comment]
ysrinu that is a good point. I initially separated the logic, but had planned not to use an include file due to the simplicity of the project, but it is a good time to practice the 'best practice'. Also not bad for me to run through the process of redeploying to my production server after making changes.

**Comment by drumph33** [comment]
Ah interesting discussion, thanks for bringing this up.  I went back in and changed mine around so that the php logic is in a different file.  Good practice!

**Comment by susanBuck** [comment]
Good comments by all in this thread. I'm going to close this out as resolved but feel free to continue the discussion!
